src/base/par_iter.rs

```diff
@@ -0,0 +1,156 @@
+//! this module implements parallelators to make matrices work with
+//! the rayon crate seamlessly
+
+use crate::{
+    iter::{ColumnIter, ColumnIterMut},
+    Dim, Matrix, MatrixSlice, MatrixSliceMut, RawStorage, RawStorageMut, U1,
+};
+use core::fmt::Debug;
+use rayon::{iter::plumbing::bridge, prelude::*};
+
+/// A rayon parallel iterator over the colums of a matrix
+pub struct ParColumnIter<'a, T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols>> {
+    mat: &'a Matrix<T, R, Cols, S>,
+}
+
+impl<'a, T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols>> ParColumnIter<'a, T, R, Cols, S> {
+    /// create a new parallel iterator for the given matrix
+    fn new(matrix: &'a Matrix<T, R, Cols, S>) -> Self {
+        Self { mat: matrix }
+    }
+}
+
+impl<'a, T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols>> ParallelIterator
+    for ParColumnIter<'a, T, R, Cols, S>
+where
+    T: Sync + Send + Clone + Debug + PartialEq + 'static,
+    S: Sync,
+{
+    type Item = MatrixSlice<'a, T, R, U1, S::RStride, S::CStride>;
+
+    fn drive_unindexed<Consumer>(self, consumer: Consumer) -> Consumer::Result
+    where
+        Consumer: rayon::iter::plumbing::UnindexedConsumer<Self::Item>,
+    {
+        bridge(self, consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.mat.ncols())
+    }
+}
+
+impl<'a, T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols>> IndexedParallelIterator
+    for ParColumnIter<'a, T, R, Cols, S>
+where
+    T: Send + Sync + Clone + Debug + PartialEq + 'static,
```

Comment:
What's the purpose of the `PartialEq` and `Debug` bounds? Maybe just require `Scalar + Send + Sync`? (applies throughout)

Reply:
Oh heck, this was a classic case of what Matt Godbolt calls Compiler-Error-Driven-Development XD. What I should have said was `Scalar` instead of the Confusing `Clone + Debug + PartialEq`. This took me a minute to figure out. I'll fix it.